designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,627 @@
+
+```js
+const fs = require("fs");
+
+exports.config = {
+    files: ["*.js"],
+    ignores: fs.readFileSync(".eslintignore", "utf8").split("\n")
+};
+```
+
+### Replacing `--ext`
+
+The `--ext` flag is currently used to pass in one or more file extensions that ESLint should search for when a directory without a glob pattern is passed on the command line, such as:
+
+```bash
+eslint src/ --ext .js,.jsx
+```
+
+This curently searches all subdirectories of `src/` for files with extensions matching `.js` or `.jsx`.
+
+This proposal removes `--ext` by allowing the same information to be passed in a config. For example, the following config achieves the same result:
+
+```js
+const fs = require("fs");
+
+exports.config = {
+    files: ["*.js", "*.jsx"],
+};
+```
+
+ESLint could then be run with this command:
+
+```bash
+eslint src/
+```
+
+When evaluating the `files` array in the config, ESLint will end up searching for `src/**/*.js` and `src/**/*.jsx`. (More information about file resolution is included later this proposal.)
+
+### The `@eslint/config` Package
+
+Because some of the operations ESLint currently do are quite complicated, this design includes a utility package called `@eslint/config` that users can use for common tasks. This package contains the following utilities:
+
+1. `RuleLoader` - a class that aids in loading rules from a directory.
+
+The `@eslint/config` package is intended to add back functionality that would be removed from ESLint with this design.
+
+**Note:** It is not required to use this package with `eslint.config.js`; this is an optional tool that users may choose to use for convenience.
+
+#### The `RuleLoader` Class
+
+The purpose of the `RuleLoader` class is to aid in loading rules to replace `--rulesdir` functionality and has this form:
+
+```js
+class RuleLoader {
+
+    // create a new instance
+    constructor() {}
+
+    // mimic `--rulesdir` loading
+    loadRulesFromDirectory(directory) {}
+}
+```
+
+In order to recreate the functionality of `--rulesdir`, a user would need to create a new entry in `ruledefs` and then specify the rules from a directory, such as:
+
+```js
+const { RuleLoader } = require("@eslint/config");
+const ruleLoader = new RuleLoader();
+
+exports.config = {
+    ruledefs: {
+        custom: ruleLoader.loadFromDirectory("./custom-rules")
+    },
+    rules: {
+        "custom/my-rule": "error"
+    }
+};
+```
+
+The `RuleLoader#loadFromDirectory()` method returns an object where the keys are the rule IDs (based on the filenames found in the directory) and the values are the rule objects. Unlike today, rules loaded from a local directory must have a namespace just like plugin rules (`custom` in this example).
+
+### Advanced Configs
+
+Some users may need information from ESLint to determine the correct configuration to use. To allow for that, `exports.config` may also be a function that returns an object, such as:
```

Comment:
1. The function must return an object or an array.
2. No.
3. No.

It helps to think of the function like an AWS Lambda function. Inside of that function you can do whatever you want, but returning a function from that function won't result in a second function call.

I'll add in some details about this.

Reply:
Does this mean that in order to use a shareable config that exports a function, an end user would also be required to make their config a function (so they could call the shareable config with the right arguments)?

I'm concerned that this would cause some problems. For example:

* A shareable config would be required to make a breaking change to convert their config from an object/array to a function.
* An end user might be confused about why some shareable configs can be extended with `extends: [require('eslint-config-foo')]`, and other shareable configs need to be extended with `extends: [require('eslint-config-foo')(context)]`.

Could you elaborate on what we gain by disallowing functions in `extends`? In my opinion, it seems better to allow a function anywhere we allow an object config.

Reply:
> Does this mean that in order to use a shareable config that exports a function, an end user would also be required to make their config a function (so they could call the shareable config with the right arguments)?

That was my intention, yes. However, you make some good points and have convinced me that it makes sense to allow a function anywhere a config object is currently allowed. I'll work on adding that into this design this week.